In [1]:
import pandas as pd

# Load in 2012 US Voter Data

This is 2012 population and voter registration data 
from the US Census Bureau, 
broken down by state and by age group.

In [2]:
df = pd.read_csv("2012_US_Voter_data.csv")

# Initial EDA

In [3]:
df.head()

State       Age  Total Population  Citizen Population  Registered Voters  \
0  Alabama  18 to 24            439000              428000             212000   
1  Alabama  25 to 34            576000              535000             359000   
2  Alabama  35 to 44            615000              582000             410000   
3  Alabama  45 to 64           1297000             1275000            1051000   
4  Alabama       65+            667000              660000             523000   

   Confirmed Voters  
0            155000  
1            271000  
2            330000  
3            939000  
4            459000

In [4]:
df.tail()

State       Age  Total Population  Citizen Population  \
250  Wyoming  18 to 24             56000               55000   
251  Wyoming  25 to 34             73000               71000   
252  Wyoming  35 to 44             68000               66000   
253  Wyoming  45 to 64            155000              154000   
254  Wyoming       65+             74000               73000   

     Registered Voters  Confirmed Voters  
250              21000             18000  
251              44000             39000  
252              41000             36000  
253             101000             95000  
254              61000             59000

In [5]:
df.sample(10)

State       Age  Total Population  Citizen Population  \
60                  Idaho  18 to 24            130000              121000   
186                Oregon  25 to 34            461000              399000   
99                  Maine       65+            219000              215000   
107         Massachusetts  35 to 44            888000              783000   
41   District Of Columbia  25 to 34            154000              133000   
189                Oregon       65+            616000              604000   
8                  Alaska  45 to 64            186000              182000   
81                 Kansas  25 to 34            406000              359000   
78                   Iowa  45 to 64            828000              812000   
23             California  45 to 64           9356000             7827000   

     Registered Voters  Confirmed Voters  
60               51000             39000  
186             256000            215000  
99              187000            170000  
107             580000            530000  
41              114000            102000  
189             513000            491000  
8               144000            122000  
81              236000            172000  
78              652000            586000  
23             5459000           4926000

In [6]:
df.columns

Index(['State', 'Age', 'Total Population', 'Citizen Population',
       'Registered Voters', 'Confirmed Voters'],
      dtype='object')

In [7]:
df. describe()

Total Population  Citizen Population  Registered Voters  \
count      2.550000e+02        2.550000e+02       2.550000e+02   
mean       9.225451e+05        8.434706e+05       6.006000e+05   
std        1.157013e+06        1.007374e+06       7.203145e+05   
min        5.600000e+04        5.500000e+04       2.100000e+04   
25%        2.385000e+05        2.240000e+05       1.465000e+05   
50%        5.680000e+05        5.400000e+05       3.710000e+05   
75%        1.159000e+06        1.029000e+06       7.655000e+05   
max        9.356000e+06        7.827000e+06       5.459000e+06   

       Confirmed Voters  
count      2.550000e+02  
mean       5.213804e+05  
std        6.413869e+05  
min        1.800000e+04  
25%        1.285000e+05  
50%        3.030000e+05  
75%        6.640000e+05  
max        4.926000e+06

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 255 entries, 0 to 254
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   State               255 non-null    object
 1   Age                 255 non-null    object
 2   Total Population    255 non-null    int64 
 3   Citizen Population  255 non-null    int64 
 4   Registered Voters   255 non-null    int64 
 5   Confirmed Voters    255 non-null    int64 
dtypes: int64(4), object(2)
memory usage: 12.1+ KB


# Questions to think about for the US_Voter_2012 dataset

some of the questions that I'd like to
explore here are things like 

	• how does voter turnout compare across states?

	• Are certain age ranges more or less likely to vote in general?

	• And do these trends hold true nationwide, 

or are there state specific insights that we can uncover here?

# Lets look into the "Total Population", "Citizen Population", "Registered Voters", "Confirmed Voters"

Let's cut our data by state as rows 

and pull in some of the metrics like 'Citizen Population' and 'Confirmed Voters'.

For now, let's use citizen population since only citizens can vote and confirmed voters and start there
and we'll go ahead and format both those as numbers with 1000 separator and no decimal points.
There we go.
And finally, I'll sort my states descending by the citizen population.

So as expected, we see big states like California, Texas, Florida and New York at the top of our
list.

And then down at the bottom, obviously, are less densely populated states like the Dakotas and Alaska,
Vermont and Wyoming.

In [9]:
df.groupby("State")['Citizen Population','Confirmed Voters'].sum().reset_index().sort_values('Confirmed Voters', ascending=False)

/tmp/ipykernel_30840/1432508034.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df.groupby("State")['Citizen Population','Confirmed Voters'].sum().reset_index().sort_values('Confirmed Voters', ascending=False)


State  Citizen Population  Confirmed Voters
4             California            23418000          13463000
43                 Texas            16062000           8643000
9                Florida            13326000           8107000
32              New York            13081000           7676000
38          Pennsylvania             9451000           5824000
13              Illinois             8831000           5428000
35                  Ohio             8550000           5395000
22              Michigan             7228000           4831000
33        North Carolina             6711000           4624000
10               Georgia             6738000           4168000
46              Virginia             5646000           3778000
30            New Jersey             5929000           3670000
21         Massachusetts             4773000           3382000
47            Washington             4833000           3172000
49             Wisconsin             4247000           3128000
23             Minnesota             3903000           2859000
25              Missouri             4409000           2818000
14               Indiana             4724000           2801000
20              Maryland             4006000           2610000
42             Tennessee             4678000           2606000
5               Colorado             3543000           2495000
2                Arizona             4315000           2412000
40        South Carolina             3381000           2187000
0                Alabama             3480000           2154000
18             Louisiana             3239000           2149000
37                Oregon             2807000           1897000
17              Kentucky             3193000           1895000
24           Mississippi             2130000           1588000
6            Connecticut             2500000           1569000
15                  Iowa             2232000           1548000
36              Oklahoma             2734000           1431000
16                Kansas             1974000           1249000
3               Arkansas             2110000           1124000
28                Nevada             1808000           1048000
44                  Utah             1793000           1022000
31            New Mexico             1426000            878000
27              Nebraska             1297000            799000
19                 Maine             1020000            699000
48         West Virginia             1443000            689000
29         New Hampshire              992000            688000
12                 Idaho             1064000            680000
26               Montana              754000            495000
11                Hawaii              930000            481000
39          Rhode Island              751000            469000
7               Delaware              642000            430000
41          South Dakota              606000            371000
8   District Of Columbia              461000            351000
34          North Dakota              515000            328000
45               Vermont              487000            307000
1                 Alaska              495000            289000
50               Wyoming              419000            247000


# Create a calculated field '% Voter Population'

From here, I want to add a calculated field, because what I'm interested in looking at isn't just

the number of citizens and the number of confirmed voters, which is interesting but not entirely surprising.

What I want to do is actually calculate a ratio so that I can see out of the citizen population what

percentage of those actually became confirmed voters.

${\% Voter Population = \frac{(Confirmed Voters)}{Citizen Population}}$ 

## Lets find the state with the highest '% Voter Population'

at the state level

In [10]:
df_state_level = df.groupby("State")['Citizen Population','Confirmed Voters'].sum().reset_index()

df_state_level.head(5)

/tmp/ipykernel_30840/751649940.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df_state_level = df.groupby("State")['Citizen Population','Confirmed Voters'].sum().reset_index()


State  Citizen Population  Confirmed Voters
0     Alabama             3480000           2154000
1      Alaska              495000            289000
2     Arizona             4315000           2412000
3    Arkansas             2110000           1124000
4  California            23418000          13463000

In [11]:
df_state_level['% Voter Population - State_Level'] = round((df_state_level['Confirmed Voters'] / df_state_level['Citizen Population'] * 100), 2)

In [12]:
df_state_level = df_state_level.sort_values(by=['% Voter Population - State_Level'], ascending=False)

## the Top 5 states with highest '% Voter Population'

We see District of Columbia or D.C. at the top of the list with a voter population percentage of 76.1%.

That makes sense.

It's the nation's capital.

It's the political heart of the country.

But then you also have states like Mississippi and Wisconsin, which, you know, may come as a surprise
to some.

In [13]:
df_state_level.head()

State  Citizen Population  Confirmed Voters  \
8   District Of Columbia              461000            351000   
24           Mississippi             2130000           1588000   
49             Wisconsin             4247000           3128000   
23             Minnesota             3903000           2859000   
21         Massachusetts             4773000           3382000   

    % Voter Population - State_Level  
8                              76.14  
24                             74.55  
49                             73.65  
23                             73.25  
21                             70.86

## the Bottom 5 states with lowest '% Voter Population'

Now, scrolling down on the other end of the spectrum, we see West Virginia, Hawaii, Oklahoma and
Arkansas.

Now, this isn't necessarily the time or place to dig much deeper into why these trends are the way
they are.

So what actually causes the West Virginia population to be less likely to convert into voters compared
to Massachusetts or Colorado, for instance?

In [14]:
df_state_level.tail()

State  Citizen Population  Confirmed Voters  \
43          Texas            16062000           8643000   
3        Arkansas             2110000           1124000   
36       Oklahoma             2734000           1431000   
11         Hawaii              930000            481000   
48  West Virginia             1443000            689000   

    % Voter Population - State_Level  
43                             53.81  
3                              53.27  
36                             52.34  
11                             51.72  
48                             47.75

So maybe things like 

	• income per capita 

	• or education level 

	• or additional demographics 

might help to provide some additional context 

and give you tools to better understand the why or the driving forces behind

these voter rates that we're seeing here.

In [15]:
df_state_level

State  Citizen Population  Confirmed Voters  \
8   District Of Columbia              461000            351000   
24           Mississippi             2130000           1588000   
49             Wisconsin             4247000           3128000   
23             Minnesota             3903000           2859000   
21         Massachusetts             4773000           3382000   
5               Colorado             3543000           2495000   
15                  Iowa             2232000           1548000   
29         New Hampshire              992000            688000   
33        North Carolina             6711000           4624000   
19                 Maine             1020000            699000   
37                Oregon             2807000           1897000   
7               Delaware              642000            430000   
46              Virginia             5646000           3778000   
22              Michigan             7228000           4831000   
18             Louisiana             3239000           2149000   
26               Montana              754000            495000   
47            Washington             4833000           3172000   
20              Maryland             4006000           2610000   
40        South Carolina             3381000           2187000   
12                 Idaho             1064000            680000   
25              Missouri             4409000           2818000   
34          North Dakota              515000            328000   
16                Kansas             1974000           1249000   
35                  Ohio             8550000           5395000   
45               Vermont              487000            307000   
6            Connecticut             2500000           1569000   
39          Rhode Island              751000            469000   
30            New Jersey             5929000           3670000   
0                Alabama             3480000           2154000   
10               Georgia             6738000           4168000   
38          Pennsylvania             9451000           5824000   
27              Nebraska             1297000            799000   
31            New Mexico             1426000            878000   
13              Illinois             8831000           5428000   
41          South Dakota              606000            371000   
9                Florida            13326000           8107000   
17              Kentucky             3193000           1895000   
14               Indiana             4724000           2801000   
50               Wyoming              419000            247000   
32              New York            13081000           7676000   
1                 Alaska              495000            289000   
28                Nevada             1808000           1048000   
4             California            23418000          13463000   
44                  Utah             1793000           1022000   
2                Arizona             4315000           2412000   
42             Tennessee             4678000           2606000   
43                 Texas            16062000           8643000   
3               Arkansas             2110000           1124000   
36              Oklahoma             2734000           1431000   
11                Hawaii              930000            481000   
48         West Virginia             1443000            689000   

    % Voter Population - State_Level  
8                              76.14  
24                             74.55  
49                             73.65  
23                             73.25  
21                             70.86  
5                              70.42  
15                             69.35  
29                             69.35  
33                             68.90  
19                             68.53  
37                             67.58  
7                              66.98  
46                             66.91  
22                             66.84  
1

# Lets look at the age bucket distribution overall

Now, the next thing that 
I'd like to explore a bit more here to kick things off is 
the composition of the voter population.

In [29]:
## Let's look at Voting per Age bucket
df_age_level = df.groupby("Age")['Citizen Population','Confirmed Voters'].sum().reset_index()
# new_df = new_df.sort_values(by=['Citizen Population'], ascending=False)
df_age_level


/tmp/ipykernel_30840/2997487765.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df_age_level = df.groupby("Age")['Citizen Population','Confirmed Voters'].sum().reset_index()


Age  Citizen Population  Confirmed Voters
0  18 to 24            27537000          11351000
1  25 to 34            35474000          18978000
2  35 to 44            34266000          20965000
3  45 to 64            76641000          52013000
4       65+            41167000          29645000

### lets' add calculated field-  '% Voter Population - Age_Level'

This '% Voter Population - Age_Level' is calculated per Age Bucket over all 50 states

Now, what this tells me is that among all confirmed voters in this data set, 
so in the year 2012,

I know that the largest share of those voters fall into the 45 to 64 range, 
followed by the 65 plus range at 22%.

And on the other end, 
you've got your smallest proportion or share of voters which fall into 18 to 24.

In [30]:

df_age_level['% Voter Population - Age_Level_overall'] = round((df_age_level['Confirmed Voters'] / df_age_level['Citizen Population'] * 100),2)
df_age_level

Age  Citizen Population  Confirmed Voters  \
0  18 to 24            27537000          11351000   
1  25 to 34            35474000          18978000   
2  35 to 44            34266000          20965000   
3  45 to 64            76641000          52013000   
4       65+            41167000          29645000   

   % Voter Population - Age_Level_overall  
0                                   41.22  
1                                   53.50  
2                                   61.18  
3                                   67.87  
4                                   72.01

In [31]:
# Let's add another calculated field which a % of column total

In [32]:
total_confirmed_voters = df_age_level['Confirmed Voters'].sum()
total_confirmed_voters

132952000

In [33]:
df_age_level['% of Total Confirmed Voters'] = round((df_age_level['Confirmed Voters'] / total_confirmed_voters *100), 2)

In [34]:
df_age_level.columns

Index(['Age', 'Citizen Population', 'Confirmed Voters',
       '% Voter Population - Age_Level_overall',
       '% of Total Confirmed Voters'],
      dtype='object')

In [37]:
df_age_level = df_age_level[['Age', 
                             'Citizen Population', 
                             'Confirmed Voters',
                             '% of Total Confirmed Voters',
                             '% Voter Population - Age_Level_overall',
       ]]
df_age_level

Age  Citizen Population  Confirmed Voters  \
0  18 to 24            27537000          11351000   
1  25 to 34            35474000          18978000   
2  35 to 44            34266000          20965000   
3  45 to 64            76641000          52013000   
4       65+            41167000          29645000   

   % of Total Confirmed Voters  % Voter Population - Age_Level_overall  
0                         8.54                                   41.22  
1                        14.27                                   53.50  
2                        15.77                                   61.18  
3                        39.12                                   67.87  
4                        22.30                                   72.01

Now, what this tells me is that among all confirmed voters in this data set, 
so in the year 2012,

I know that the largest share of those voters fall into the 45 to 64 range, 
followed by the 65 plus range at 22%.

And on the other end, 

you've got your smallest proportion or share of voters which fall into 18 to 24.

So the youngest voters make up only 8.54% of the total number of confirmed voters nationwide.

# Lets calculate the '% of Total Confirmed Voters per State'

### lets' add calculated field-  '% Voter Population - Age_Level per State'

This '% Voter Population - Age_Level' is calculated per Age Bucket over all 50 states

In [38]:
df['State'].unique()

array(['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'Delaware', 'District Of Columbia',
       'Florida', 'Georgia', 'Hawaii', 'Idaho', 'Illinois', 'Indiana',
       'Iowa', 'Kansas', 'Kentucky', 'Louisiana', 'Maine', 'Maryland',
       'Massachusetts', 'Michigan', 'Minnesota', 'Mississippi',
       'Missouri', 'Montana', 'Nebraska', 'Nevada', 'New Hampshire',
       'New Jersey', 'New Mexico', 'New York', 'North Carolina',
       'North Dakota', 'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania',
       'Rhode Island', 'South Carolina', 'South Dakota', 'Tennessee',
       'Texas', 'Utah', 'Vermont', 'Virginia', 'Washington',
       'West Virginia', 'Wisconsin', 'Wyoming'], dtype=object)

In [39]:
# for Alabama

test_state = df['State'].unique()[0]
test_state

'Alabama'

In [40]:
test_state_df = df[df['State'] == test_state]
test_state_df

State       Age  Total Population  Citizen Population  Registered Voters  \
0  Alabama  18 to 24            439000              428000             212000   
1  Alabama  25 to 34            576000              535000             359000   
2  Alabama  35 to 44            615000              582000             410000   
3  Alabama  45 to 64           1297000             1275000            1051000   
4  Alabama       65+            667000              660000             523000   

   Confirmed Voters  
0            155000  
1            271000  
2            330000  
3            939000  
4            459000

In [41]:
test_state_df['% of Total Confirmed Voters per State'] = round((test_state_df['Confirmed Voters']/test_state_df['Confirmed Voters'].sum()*100),2)

/tmp/ipykernel_30840/2378482979.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test_state_df['% of Total Confirmed Voters per State'] = round((test_state_df['Confirmed Voters']/test_state_df['Confirmed Voters'].sum()*100),2)


In [42]:
test_state_df

State       Age  Total Population  Citizen Population  Registered Voters  \
0  Alabama  18 to 24            439000              428000             212000   
1  Alabama  25 to 34            576000              535000             359000   
2  Alabama  35 to 44            615000              582000             410000   
3  Alabama  45 to 64           1297000             1275000            1051000   
4  Alabama       65+            667000              660000             523000   

   Confirmed Voters  % of Total Confirmed Voters per State  
0            155000                                   7.20  
1            271000                                  12.58  
2            330000                                  15.32  
3            939000                                  43.59  
4            459000                                  21.31

In [54]:
# state_list = df['State'].unique()
# state_list[1:]

array(['Alaska', 'Arizona', 'Arkansas', 'California', 'Colorado',
       'Connecticut', 'Delaware', 'District Of Columbia', 'Florida',
       'Georgia', 'Hawaii', 'Idaho', 'Illinois', 'Indiana', 'Iowa',
       'Kansas', 'Kentucky', 'Louisiana', 'Maine', 'Maryland',
       'Massachusetts', 'Michigan', 'Minnesota', 'Mississippi',
       'Missouri', 'Montana', 'Nebraska', 'Nevada', 'New Hampshire',
       'New Jersey', 'New Mexico', 'New York', 'North Carolina',
       'North Dakota', 'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania',
       'Rhode Island', 'South Carolina', 'South Dakota', 'Tennessee',
       'Texas', 'Utah', 'Vermont', 'Virginia', 'Washington',
       'West Virginia', 'Wisconsin', 'Wyoming'], dtype=object)

In [59]:
pd.set_option('display.max_columns', None)

In [65]:
state_list = df['State'].unique()
df_list = []

for state in state_list:
    # print(state)
    a = df[df['State'] == state]
    a['% of Total Confirmed Voters per State'] = round((a['Confirmed Voters']/a['Confirmed Voters'].sum()*100),2)
    # print(a.head())
    df_list.append(a)

/tmp/ipykernel_30840/1932299257.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  a['% of Total Confirmed Voters per State'] = round((a['Confirmed Voters']/a['Confirmed Voters'].sum()*100),2)
/tmp/ipykernel_30840/1932299257.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  a['% of Total Confirmed Voters per State'] = round((a['Confirmed Voters']/a['Confirmed Voters'].sum()*100),2)
/tmp/ipykernel_30840/1932299257.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataF

In [66]:
df_list[6]

State       Age  Total Population  Citizen Population  \
30  Connecticut  18 to 24            333000              308000   
31  Connecticut  25 to 34            402000              331000   
32  Connecticut  35 to 44            441000              382000   
33  Connecticut  45 to 64           1047000              993000   
34  Connecticut       65+            503000              486000   

    Registered Voters  Confirmed Voters  % of Total Confirmed Voters per State  
30             145000            123000                                   7.84  
31             210000            177000                                  11.28  
32             255000            223000                                  14.21  
33             768000            695000                                  44.30  
34             383000            351000                                  22.37

In [77]:
# next concat list of df into one big dataframe
df_age_level_concat = pd.concat(df_list)

In [78]:
df_age_level_concat

State       Age  Total Population  Citizen Population  \
0    Alabama  18 to 24            439000              428000   
1    Alabama  25 to 34            576000              535000   
2    Alabama  35 to 44            615000              582000   
3    Alabama  45 to 64           1297000             1275000   
4    Alabama       65+            667000              660000   
..       ...       ...               ...                 ...   
250  Wyoming  18 to 24             56000               55000   
251  Wyoming  25 to 34             73000               71000   
252  Wyoming  35 to 44             68000               66000   
253  Wyoming  45 to 64            155000              154000   
254  Wyoming       65+             74000               73000   

     Registered Voters  Confirmed Voters  \
0               212000            155000   
1               359000            271000   
2               410000            330000   
3              1051000            939000   
4               523000            459000   
..                 ...               ...   
250              21000             18000   
251              44000             39000   
252              41000             36000   
253             101000             95000   
254              61000             59000   

     % of Total Confirmed Voters per State  
0                                     7.20  
1                                    12.58  
2                                    15.32  
3                                    43.59  
4                                    21.31  
..                                     ...  
250                                   7.29  
251                                  15.79  
252                                  14.57  
253                                  38.46  
254                                  23.89  

[255 rows x 7 columns]

In [70]:
# now merge with state level % vote calculations
df_state_level.head(10)

State  Citizen Population  Confirmed Voters  \
8   District Of Columbia              461000            351000   
24           Mississippi             2130000           1588000   
49             Wisconsin             4247000           3128000   
23             Minnesota             3903000           2859000   
21         Massachusetts             4773000           3382000   
5               Colorado             3543000           2495000   
15                  Iowa             2232000           1548000   
29         New Hampshire              992000            688000   
33        North Carolina             6711000           4624000   
19                 Maine             1020000            699000   

    % Voter Population - State_Level  
8                              76.14  
24                             74.55  
49                             73.65  
23                             73.25  
21                             70.86  
5                              70.42  
15                             69.35  
29                             69.35  
33                             68.90  
19                             68.53

In [72]:
df_state_level_premerge = df_state_level[['State', '% Voter Population - State_Level']]

In [73]:
df_state_level_premerge.head(10)

State  % Voter Population - State_Level
8   District Of Columbia                             76.14
24           Mississippi                             74.55
49             Wisconsin                             73.65
23             Minnesota                             73.25
21         Massachusetts                             70.86
5               Colorado                             70.42
15                  Iowa                             69.35
29         New Hampshire                             69.35
33        North Carolina                             68.90
19                 Maine                             68.53

In [80]:
# merge on 'State' Column

merged_df = pd.merge(df_age_level_concat, df_state_level_premerge, on='State')
merged_df

State       Age  Total Population  Citizen Population  \
0    Alabama  18 to 24            439000              428000   
1    Alabama  25 to 34            576000              535000   
2    Alabama  35 to 44            615000              582000   
3    Alabama  45 to 64           1297000             1275000   
4    Alabama       65+            667000              660000   
..       ...       ...               ...                 ...   
250  Wyoming  18 to 24             56000               55000   
251  Wyoming  25 to 34             73000               71000   
252  Wyoming  35 to 44             68000               66000   
253  Wyoming  45 to 64            155000              154000   
254  Wyoming       65+             74000               73000   

     Registered Voters  Confirmed Voters  \
0               212000            155000   
1               359000            271000   
2               410000            330000   
3              1051000            939000   
4               523000            459000   
..                 ...               ...   
250              21000             18000   
251              44000             39000   
252              41000             36000   
253             101000             95000   
254              61000             59000   

     % of Total Confirmed Voters per State  % Voter Population - State_Level  
0                                     7.20                             61.90  
1                                    12.58                             61.90  
2                                    15.32                             61.90  
3                                    43.59                             61.90  
4                                    21.31                             61.90  
..                                     ...                               ...  
250                                   7.29                             58.95  
251                                  15.79                             58.95  
252                                  14.57                             58.95  
253                                  38.46                             58.95  
254                                  23.89                             58.95  

[255 rows x 8 columns]

Notes for Pandas Pivot table

https://www.youtube.com/watch?v=5yFox2cReTw

how to sort in pandas pivot table

https://www.statology.org/pandas-pivot-table-sort/

Notes to add a calculated field to a pandas pivot table

https://stackoverflow.com/questions/70281086/how-to-add-calculated-fields-in-pandas-pivot-table

In [ ]:
# this is the visual pivot of the previous dateframe 'df_state_level'
 
pivot_StateLevel_VoterPopPer = pd.pivot_table(data=df, index=['State','Age'], values=['Citizen Population', 'Confirmed Voters'], aggfunc='sum')
p = pivot_StateLevel_VoterPopPer.sort_values(by=['Citizen Population'], ascending=False).reset_index()
p['% Voter Population'] = p['Confirmed Voters'] / p['Citizen Population']
pd.pivot_table(data=p, index=['State','Age']).sort_values(by=['State','% Voter Population'], ascending=[False,False])